### PR TITLE
CP-16936: Implement GUI for viewing the PVS farms and PVS servers (read-only)

### PR DESCRIPTION
**Pull request by Frezzle**
Added new dialog for viewing PVS farms and servers.

Signed-off-by: Frezzle <frederico.mazzone@citrix.com>

### CODE REVIEW

File: XenAdmin/Dialogs/PvsFarmDialog.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        /// Creates a dialog for viewing the PVS farms and PVS servers on a particular connection.
+        /// </summary>
+        /// <param name="_connection">May not be null.</param>
+        public PvsFarmDialog(IXenConnection _connection)
+        {
+            System.Diagnostics.Trace.Assert(_connection != null);
+            connection = _connection;
+
+            InitializeComponent();
+            Text += string.Format(" - '{0}'", this.connection.Name);
```
Comment (MihaelaStoica) [inline XenAdmin/Dialogs/PvsFarmDialog.cs]: We should put the string in the resources (Messages.resx). Also, shouldn't this be something like "PVS Farms - {0}"?

File: XenAdmin/Dialogs/PvsFarmDialog.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        private void Rebuild()
+        {
+            Program.AssertOnEventThread();
+
+            try
+            {
+                gridView.SuspendLayout();
+                gridView.Rows.Clear();
+
+                PVS_farm[] pvsFarmsArray = connection.Cache.PVS_farms;
```
Comment (MihaelaStoica) [inline XenAdmin/Dialogs/PvsFarmDialog.cs]: We can use Sort() function here, which will sort the XenObjets by Name property (which we will need to define). E.g.
`var pvsFarms = new List<PVS_farm>(Connection.Cache.PVS_farms);`
`pvsFarms .Sort();`

File: XenAdmin/Dialogs/PvsFarmDialog.resx [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+  <data name="$this.AutoScaleDimensions" type="System.Drawing.SizeF, System.Drawing">
+    <value>96, 96</value>
+  </data>
+  <data name="$this.ClientSize" type="System.Drawing.Size, System.Drawing">
+    <value>618, 361</value>
+  </data>
+  <data name="$this.Font" type="System.Drawing.Font, System.Drawing">
+    <value>Tahoma, 8pt</value>
+  </data>
+  <data name="$this.Icon" type="System.Drawing.Icon, System.Drawing" mimetype="application/x-microsoft.net.object.bytearray.base64">
```
Comment (MihaelaStoica) [inline XenAdmin/Dialogs/PvsFarmDialog.resx]: The dialog shouldn't have the icon embedded in the resource, it should inherit application's icon

File: XenAdmin/Controls/DataGridViewEx/CollapsingPvsFarmServerDataGridViewRow.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+                SetNoIcon();
+
+                ipAddressesCell.Value = string.Join(", ", server.addresses);
+                firstPortCell.Value = server.first_port;
+                lastPortCell.Value = server.last_port;
+            }
+        }
+
+        public int ExpansionCellIndex
```
Comment (MihaelaStoica) [inline XenAdmin/Controls/DataGridViewEx/CollapsingPvsFarmServerDataGridViewRow.cs]: Are these public properties used anywhere?

Comment (Frezzle) [inline XenAdmin/Controls/DataGridViewEx/CollapsingPvsFarmServerDataGridViewRow.cs]: UnderlyingFarm and IsFarmRow are used but the others aren't. I will remove the index-related ones, but should I also remove the UnderlyingServer and IsServerRow, or should I keep them for completeness/future-use (as their "counterparts" are currently in use)?

File: XenAdmin/Dialogs/PvsFarmDialog.cs (hunk 1 above)
Comment (Frezzle) [inline XenAdmin/Dialogs/PvsFarmDialog.cs]: The Text property is initialized to "PVS Farms" by the designer and then here we concatenate " - '{0}'". I'll add a string resource and change this to an assignment similar to:
Text = string.Format(Messages.PVS_FARMS_DIALOG_TITLE, this.connection.Name);

Comment (stephen-turner) [inline XenAdmin/Dialogs/PvsFarmDialog.cs]: I'm not commenting on this specific case, but concatenation of strings generally doesn't work well when localising: other languages may have the parts in a different order, or use different separators, or different parts of speech in different contexts, or... So it's generally better to specify the whole string.

File: XenAdmin/Controls/DataGridViewEx/CollapsingPvsFarmServerDataGridViewRow.cs (hunk 4 above)
Comment (MihaelaStoica) [inline XenAdmin/Controls/DataGridViewEx/CollapsingPvsFarmServerDataGridViewRow.cs]: Let's keep UnderlyingServer and IsServerRow, for completeness

**Comment by Frezzle** [comment]
@MihaelaStoica All changes made :+1:

**Comment by MihaelaStoica** [comment]
Looks good. Merging

**Pull request merged**